plugin-server/src/main/ingestion-queues/session-recording-v2/sessions/recorder.ts

```diff
@@ -1,44 +1,43 @@
-import { Writable } from 'stream'
+import { Readable } from 'stream'
+import { createGzip } from 'zlib'
 
 import { ParsedMessageData } from '../kafka/types'
 
-interface WriteResult {
+export interface EndResult {
+    stream: Readable
     eventCount: number
-    bytesWritten: number
 }
 
 export class SessionRecorder {
-    private chunks: string[] = []
-    private size: number = 0
+    private readonly gzip: ReturnType<typeof createGzip>
+    private eventCount: number = 0
+    private rawBytesWritten: number = 0
+
+    constructor() {
+        this.gzip = createGzip()
+    }
 
     public recordMessage(message: ParsedMessageData): number {
-        let bytesWritten = 0
+        let rawBytesWritten = 0
 
         Object.entries(message.eventsByWindowId).forEach(([windowId, events]) => {
             events.forEach((event) => {
                 const serializedLine = JSON.stringify([windowId, event]) + '\n'
-                this.chunks.push(serializedLine)
-                bytesWritten += Buffer.byteLength(serializedLine)
+                this.gzip.write(serializedLine)
+                rawBytesWritten += Buffer.byteLength(serializedLine)
+                this.eventCount++
             })
         })
 
-        this.size += bytesWritten
-        return bytesWritten
+        this.rawBytesWritten += rawBytesWritten
+        return rawBytesWritten
     }
 
-    public async write(stream: Writable): Promise<WriteResult> {
-        let eventCount = 0
-        let bytesWritten = 0
-
-        for (const chunk of this.chunks) {
-            if (!stream.write(chunk)) {
-                // Handle backpressure
-                await new Promise((resolve) => stream.once('drain', resolve))
-            }
-            eventCount++
-            bytesWritten += Buffer.byteLength(chunk)
+    public end(): EndResult {
+        this.gzip.end()
```

Comment:
Does this need to be awaited? I couldn't track the control flow but it didn't seem like we are actually waiting for the callback anywhere 🤔